util/rateLimit.go

```diff
@@ -0,0 +1,321 @@
+	default:
+	}
+	return false, fmt.Errorf("unable to consume capacity from ElasticRateLimiter")
+}
+
+// ReturnCapacity will insert new capacity on the sharedCapacity or reservedCapacity of a client.
+// if the capacity could not be returned to any channel, an error is returned
+func (erl ElasticRateLimiter) ReturnCapacity(c client, reserved bool) error {
+	erl.clientMu.RLock()
+	closed := erl.closedClients[c]
+	capChan, exists := erl.capacityByClient[c]
+	numClients := len(erl.capacityByClient)
+	erl.clientMu.RUnlock()
+	// all closed clients return their capacity to sharedCapacity
+	if closed {
+		select {
+		case erl.sharedCapacity <- capacity{}:
+			return nil
+		default:
+			return fmt.Errorf("client reservation is closed and could not return capacity to sharedCapacity")
+		}
+	}
+	if !exists {
+		return fmt.Errorf("client has no reservation")
+	}
+	if reserved {
+		select {
+		case capChan <- capacity{}:
+			return nil
+		default:
+			return fmt.Errorf("could not return capacity to client reservation")
+		}
+	}
+	// before adding to the sharedCapacity, confirm it wouldn't overprovision the sharedCapacity
+	maxShared := erl.MaxCapacity - (numClients * erl.CapacityPerReservation)
+	if len(erl.sharedCapacity) >= maxShared {
+		return fmt.Errorf("capacity return would overprovision the sharedCapacity")
+	}
+	select {
+	case erl.sharedCapacity <- capacity{}:
+		return nil
+	default:
+		return fmt.Errorf("could not return capacity to sharedCapacity")
+	}
+}
+
+// ContainsReservationFor is a lock-safe check for a client having a reservation
+func (erl ElasticRateLimiter) ContainsReservationFor(c client) bool {
+	erl.clientMu.RLock()
+	defer erl.clientMu.RUnlock()
+	_, exists := erl.capacityByClient[c]
+	return exists
+}
+
+// OpenReservation creates an entry in the ElasticRateLimiter's reservedCapacity map,
+// and optimistically transfers capacity from the sharedCapacity to the reservedCapacity
+func (erl ElasticRateLimiter) OpenReservation(c client) error {
+	erl.clientMu.Lock()
+	defer erl.clientMu.Unlock()
+	if _, exists := erl.capacityByClient[c]; exists {
+		return fmt.Errorf("client already has a reservation")
+	}
+	// guard against overprovisioning, if there is less than a reservedCapacity amount left
+	remaining := erl.MaxCapacity - (erl.CapacityPerReservation * len(erl.capacityByClient))
+	if erl.CapacityPerReservation > remaining {
+		return fmt.Errorf("not enough capacity to reserve for client: %d remaining, %d requested", remaining, erl.CapacityPerReservation)
+	}
+	// make capacity for the provided client
+	erl.capacityByClient[c] = make(chan capacity, erl.CapacityPerReservation)
+	delete(erl.closedClients, c)
+
+	// start asynchronously filling the capacity
+	// by design, the ElasticRateLimiter will overprovision capacity for a newly connected host,
+	// but will resolve the discrepancy ASAP by consuming capacity from the shared capacity which it will not return
+	for i := 0; i < erl.CapacityPerReservation; i++ {
```

Comment:
I left this in for now since over-provisioning in this case was something I actually wanted. Consider it a credit for the new reservation to immediately use capacity. Since it checks that it can afford the reservation before this point, there is not a concern over over-promising reservations. The worst that could happen is you've extended reserved capacity when all your shared capacity is unavailable. I remember we had a conversation early on about how to allow clients into the buffer even when there wasn't immediately room for them.

Happy to chat about it more tho!

Reply:
No worries, I don't feel *too* strongly about it, I'll change this to be strictly not over-provisioned.